qemu: VhostUserDevice CCW device numbers

Jakob-Naucke opened the pull request:
Add CCW (s390x) device numbers to VhostUserDevices, as is with other
device types.

Fixes: #170

Signed-off-by: Jakob Naucke <jakob.naucke@ibm.com>

Jakob-Naucke commented:
The suggested change works exactly as e.g. [`(FSDevice).QemuParams()`](https://github.com/kata-containers/govmm/blob/master/qemu/qemu.go#L421), and the giant `switch` statement was already there. Do you think any functionality should be moved to an extra function?

Jakob-Naucke commented:
See https://github.com/kata-containers/kata-containers/pull/1754 for a test.

fidencio commented:
@Jakob-Naucke, hmm., Unit tests are not exactly happy, neither is travis ci. :-/

@devimc, mind to also review this one?

fidencio commented:
LGTM, assuming happy tests.

fidencio commented:
> The suggested change works exactly as e.g. [`(FSDevice).QemuParams()`](https://github.com/kata-containers/govmm/blob/master/qemu/qemu.go#L421), and the giant `switch` statement was already there. Do you think any functionality should be moved to an extra function?

And only now I see this comment. :-p. Dunno, is there a way to explicitly ignore this check for now?

Jakob-Naucke commented:
So yes, `gocyclo` linting can be disabled at function level, which I have done in the suggested update. As a mid-term solution, we might also split up [`(VhostUserDevice).QemuParams()`](https://github.com/kata-containers/govmm/blob/7183b12b070ac483bd63c229f156615ab78237b8/qemu/qemu.go#L1279) enough to be able to remove the `nolint` again. @devimc WDYT?

Also updated commit message to stress that the QEMU parameter is only added for VhostUserFS (Net, SCSI, Blk are currently unsupported on s390x).

devimc commented:
@Jakob-Naucke thanks for this PR, IMO adding `nolint` to the function is not an option, so I suggest the following change:
* Split `(vhostuserDev VhostUserDevice) QemuParams` in several functions, all functions should return the params ready to be used (including -device or -netdev):
	* (vhostuserDev VhostUserDevice) QemuNetParams []string
	* (vhostuserDev VhostUserDevice) QemuSCSIParams []string
	* (vhostuserDev VhostUserDevice) QemuBlkParams []string
	* (vhostuserDev VhostUserDevice) QemuFSParams []string

WDYT?

Jakob-Naucke commented:
@devimc sounds almost fine, but I have one concern:
`devParams` gets a `romfile` ([source](https://github.com/kata-containers/govmm/blob/7183b12b070ac483bd63c229f156615ab78237b8/qemu/qemu.go#L1345)) regardless of `VhostUserType` and I don't think we can add it to the parameters before adding the device parameter itself. Is this something you would just do in all of the suggested functions?

devimc commented:
@Jakob-Naucke I'd check `vhostuserDev.Transport.isVirtioPCI(config) && vhostuserDev.ROMFile != ""` in every function to simplify things :disappointed:, WDYT?

Jakob-Naucke commented:
@devimc OK, sure. Some repetition but I guess huge and/or complex functions aren't better either. Will get back to you.

Jakob-Naucke commented:
Updated accordingly, also re-revendored https://github.com/kata-containers/kata-containers/pull/1754

devimc commented on qemu/qemu.go:
```diff
@@ -1275,90 +1275,149 @@ func (vhostuserDev VhostUserDevice) Valid() bool {
+	case VhostUserBlk:
+		devParams = vhostuserDev.QemuBlkParams(config)
+	case VhostUserFS:
+		devParams = vhostuserDev.QemuFSParams(config)
```
missing `default: return nil`

Jakob-Naucke replied:
I hadn't because with the `nil` check below it's not strictly necessary, but I guess it makes it a bit clearer. Updated!

Jakob-Naucke replied:
Whoops, that change didn't get staged. Is updated now.

devimc commented on qemu/qemu.go:
```diff
@@ -1275,90 +1275,149 @@ func (vhostuserDev VhostUserDevice) Valid() bool {
+		devParams = vhostuserDev.QemuFSParams(config)
+	}
+
+	if devParams != nil {
+		qemuParams = append(qemuParams, devParams...)
+	} else {
+		return nil
+	}
+
+	return qemuParams
```
This should be enough, shouldn't it?

```go
	return append(qemuParams, devParams...)
```

fidencio replied:
@devimc, nops, the append is done a few lines above if devParams is not nil.

devimc replied:
:thinking:  but we can simplify that expression (if..else are not needed), right @Jakob-Naucke  ?

Jakob-Naucke replied:
Huh, I thought I had commented already. Yes, we need to check, because `QemuWhateverParams` might return `nil` and then we should return `nil` altogether instead of the chardev.
We can go without an `else` if that's what you mean (see update), but I think we need `if`.

devimc replied:
okay this looks much better :smile:  thanks

devimc reviewed (commented):
thanks @Jakob-Naucke - I've left some comments

devimc reviewed (approved):
thanks @Jakob-Naucke

devimc commented:
one more reviewer here please

fidencio reviewed (approved):
lgtm, thanks @Jakob-Naucke!
